nvm.sh

```diff
@@ -2727,19 +2730,17 @@ nvm() {
   local DEFAULT_IFS
   DEFAULT_IFS=" $(nvm_echo t | command tr t \\t)
 "
-  if [ "${-#*e}" != "$-" ]; then
-    set +e
-    local EXIT_CODE
-    IFS="${DEFAULT_IFS}" nvm "$@"
-    EXIT_CODE="$?"
-    set -e
-    return "$EXIT_CODE"
-  elif [ "${-#*a}" != "$-" ]; then
-    set +a
+  local options="$-"
+  local o
+  for o in e a E; do
+    [ "${-#*$o}" = "$-" ] ||
+      set "+$o"
+  done
+  if [ "$options" != "$-" ]; then
     local EXIT_CODE
     IFS="${DEFAULT_IFS}" nvm "$@"
     EXIT_CODE="$?"
-    set -a
+    set "-$options"
```

Comment:
hm, i thought i tried something like this and it didn't work. have you tested this in all of dash, sh, ksh, bash, and stock zsh?

Reply:
I hadn't, but after testing just now:

- [x] `dash` -- `set` takes multiple options, but `-E` corresponds with its emacs command line editor
- [ ] `ksh` -- `set` takes multiple options, but `-E` relates to profile files, can only be specified on the original command line, and is included in `$-` by default; `set -E` and `set +E` fail
- [x] `bash` -- works here with versions 3.2+
- [x] `zsh` -- `set` takes multiple options, but `-E` controls suppression of `pushd`/`popd` output
- [x] `sh` -- not sure which implementation you test against, but my system uses Bash in `sh` mode without issues

i.e. some mitigation is needed for `ksh`; and clearing `-E` on `dash` and `zsh` is harmless but unnecessary. I'll take this into account with my next push.

Reply:
Thanks for checking - ksh is a particularly tricky one.

Reply:
I've reworked this as follows:

1. Only restore the shell options that were actually cleared, rather than calling `set -<VAL>` with the original value of `$-`. This is a safer approach with all shells, but it's required on ksh because its default options can't even be passed to `set` without errors like "ksh: set: -i: unknown option"
2. Only check for `-E` on Bash; it's the only shell that implements `errtrace`, and -E means something different on other shells.
3. For this to work reliably with a space-delimited string of options to check, clear IFS first (otherwise `"e a E"` could be interpreted as a single option). I couldn't understand why `local IFS` wasn't simply set to the preferred value, so I reimplemented this without `$DEFAULT_IFS`, but please let me know if I'm missing a pecularity of `local IFS` with ksh or zsh.